### PR TITLE
Modify reanimated props updates

### PR DESCRIPTION
**Pull request by bartlomiejbloniarz**
## Summary

This PR changes the way `performOperations` and `ReanimatedCommitHook` are synchronized. The current implementations faces 2 problems:

1. Updates that come through `performOperations` after `pleaseSkipReanimatedCommit` is called in the commit hook will be deferred until the next commit. Usually it's fine since the next commit will be triggered by the next animation frame. But if there was a singular update scheduled through Reanimated, we might not see the change for a long time. This issue is more thoroughly explained in [this issue](https://github.com/software-mansion/react-native-reanimated/issues/6245).

2. In `ReanimatedCommitMarker` there is an assumption that there can be at most one commit happening on a given thread (i.e. there can't be nested commits). This isn't true, since there can be an event listener that commits some changes in a reaction to a native mount/unmount (which is a part of the commit function). [This exact scenario was observed in the New Expensify App with `react-native-keyboard-controller`](https://github.com/Expensify/App/issues/44437#issuecomment-2250696773). At first I thought this is a mistake, but [this PR in RN](https://github.com/facebook/react-native/commit/5f0435fb852cbb8abfe6cab77784ebddf1ad38ac) seems to allow for scenarios like that.

Applied fixes:

ad 1. Now instead of resetting the skip flag in reanimated after a transaction is mounted (via MountHook), we reset the flag whenever a non-empty batch is read in `performOperations`. This ensures that we don't make an unnecessary commit, but never skip any updates.

ad 2. Now we mark reanimated commits through `ReanimatedCommitShadowNode`. This is a class derived from ShadowNode that allows us to modify the root nodes traits_ (and add our custom trait). This ensures that even if the root node gets cloned it will retain the information. We couldn't derive from `RootShadowNode` since it is a final class.

### CODE REVIEW

File: packages/react-native-reanimated/Common/cpp/Fabric/ReanimatedMountHook.cpp [hunk 1]
```diff
@@ -18,13 +17,7 @@ ReanimatedMountHook::~ReanimatedMountHook() noexcept {
...
 void ReanimatedMountHook::shadowTreeDidMount(
     RootShadowNode::Shared const &,
-    double) noexcept {
-  // When commit from React Native has finished, we reset the skip commit flag
-  // in order to allow Reanimated to commit its tree
-  if (!ReanimatedCommitMarker::isReanimatedCommit()) {
-    propsRegistry_->resetReanimatedSkipCommitFlag();
-  }
-}
+    double) noexcept {}
```
Comment (tomekzaw) [inline packages/react-native-reanimated/Common/cpp/Fabric/ReanimatedMountHook.cpp:20]: Do we still need this mount hook?

File: packages/react-native-reanimated/Common/cpp/NativeModules/NativeReanimatedModule.cpp [hunk 2]
```diff
@@ -736,7 +736,19 @@ void NativeReanimatedModule::performOperations() {
...
           }
-          return cloneShadowTreeWithNewProps(oldRootShadowNode, propsMap);
+
+          auto rootNode =
+              cloneShadowTreeWithNewProps(oldRootShadowNode, propsMap);
+
+          // Mark the commit as Reanimated commit so that we can distinguish it
+          // in ReanimatedCommitHook.
+
+          auto reaShadowNode =
```
Comment (tomekzaw) [inline packages/react-native-reanimated/Common/cpp/NativeModules/NativeReanimatedModule.cpp:746]: Let's avoid using "rea" if needed.
```suggestion
          auto markedShadowNode =
```
or
```suggestion
          auto shadowNodeWithTrait =
```

File: packages/react-native-reanimated/Common/cpp/Fabric/ReanimatedCommitHook.cpp [hunk 3]
```diff
@@ -31,9 +31,15 @@ RootShadowNode::Unshared ReanimatedCommitHook::shadowTreeWillCommit(
 #else
     RootShadowNode::Unshared const &newRootShadowNode) const noexcept {
 #endif
-  if (ReanimatedCommitMarker::isReanimatedCommit()) {
+
+  auto reaShadowNode =
+      std::reinterpret_pointer_cast<ReanimatedCommitShadowNode>(
+          newRootShadowNode);
+
+  if (reaShadowNode->isReanimatedCommit()) {
```
Comment (tomekzaw) [inline packages/react-native-reanimated/Common/cpp/Fabric/ReanimatedCommitHook.cpp:39]:
```suggestion
  if (reaShadowNode->hasReanimatedCommitTrait()) {
```

File: packages/react-native-reanimated/Common/cpp/Fabric/ReanimatedCommitShadowNode.h [hunk 4]
```diff
@@ -0,0 +1,24 @@
+#pragma once
+#ifdef RCT_NEW_ARCH_ENABLED
+
+namespace reanimated {
+
+const ShadowNodeTraits::Trait reanimatedCommitTrait{
```
Comment (tomekzaw) [inline packages/react-native-reanimated/Common/cpp/Fabric/ReanimatedCommitShadowNode.h:6]:
```suggestion
constexpr ShadowNodeTraits::Trait ReanimatedCommitTrait{
```

File: packages/react-native-reanimated/Common/cpp/Fabric/ReanimatedCommitShadowNode.h [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    ShadowNodeTraits::Trait(1 << 31)};
```
Comment (tomekzaw) [inline packages/react-native-reanimated/Common/cpp/Fabric/ReanimatedCommitShadowNode.h:7]:
```suggestion
    1 << 31};
```

File: packages/react-native-reanimated/Common/cpp/Fabric/ReanimatedCommitShadowNode.h [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+class ReanimatedCommitShadowNode : public ShadowNode {
```
Comment (tomekzaw) [inline packages/react-native-reanimated/Common/cpp/Fabric/ReanimatedCommitShadowNode.h:9]:
```suggestion
class ReanimatedTraitsEnabledShadowNode : public ShadowNode {
```

**Review by tomekzaw — CHANGES_REQUESTED** [review]

File: packages/react-native-reanimated/Common/cpp/Fabric/ReanimatedMountHook.cpp (hunk 1 above)
Comment (bartlomiejbloniarz) [reply packages/react-native-reanimated/Common/cpp/Fabric/ReanimatedMountHook.cpp:20]: no

File: packages/react-native-reanimated/Common/cpp/NativeModules/NativeReanimatedModule.cpp (hunk 2 above)
Comment (tomekzaw) [reply packages/react-native-reanimated/Common/cpp/NativeModules/NativeReanimatedModule.cpp:746]: Right, how about `ReanimatedTraitAccessEnabledShadowNode`? That's probably too long.

File: packages/react-native-reanimated/Common/cpp/Fabric/ReanimatedCommitShadowNode.h [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+namespace reanimated {
+
+// We use this trait to mark that a commit was created by reanimated.
+// Traits are copied when nodes are cloned, so this information
+// won't be lost unless someone explicitly overrides it.
+// We need this information to skip unnecessary updates in
+// the commit hook.
+// Currently RN traits go up to ten, so hopefully
+// the arbitarly chosen number 27 will be safe :)
```
Comment (tomekzaw) [inline packages/react-native-reanimated/Common/cpp/Fabric/ReanimatedCommitShadowNode.h:12]:
```suggestion
// We use this trait to mark that a commit was created by Reanimated.
// Traits are copied when nodes are cloned, so this information
// won't be lost unless someone explicitly overrides it.
// We need this information to skip unnecessary updates in
// the commit hook.
// Currently RN traits go up to 10, so hopefully
// the arbitrarily chosen number 27 will be safe :)
```

**Review by tomekzaw — APPROVED** [review]

**Pull request merged**